### PR TITLE
fix(main): fix basic auth for helm pull or push

### LINKED ISSUE #12234: helm push using zot is 403
Output of `helm version`:
version.BuildInfo{Version:"v3.12.0", GitCommit:"c9f554d75773799f72ceef38c51210f1842a1dea", GitTreeState:"clean", GoVersion:"go1.20.3"}
Output of `kubectl version`:

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 

1. login 
```
helm registry login 192.168.64.1:6000 --username admin --insecure --password admin --debug
WARNING: Using --password via the CLI is insecure. Use --password-stdin.
Login Succeeded
```
2. push 
```
helm  push openebs-3.4.1.tgz   oci://192.168.64.1:6000/charts --insecure-skip-tls-verify
Error: unexpected status from POST request to https://192.168.64.1:6000/v2/charts/openebs/blobs/uploads/: 403 Forbidden
```

so I  modify code 

```
if client.resolver == nil {
		headers := http.Header{}
		headers.Set("User-Agent", version.GetUserAgent())
		headers.Set("Authorization", "Basic YWRtaW46YWRtaW4=")
		opts := []auth.ResolverOption{auth.WithResolverHeaders(headers)}

		if client.httpClient != nil {
			opts = append(opts, auth.WithResolverClient(client.httpClient))
		}
		resolver, err := client.authorizer.ResolverWithOpts(opts...)
		if err != nil {
			return nil, err
		}
		client.resolver = resolver
	}
```

Upload was successful. I think it's an oci upload error. We need to upgrade the code to support it. If you agree with me, you can assign this issue to me.

### PR DESCRIPTION
**Pull request by cuisongliu**
closes #12234 

**What this PR does / why we need it**:

add pull and push using basic auth or bearer auth

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Review by cubxxw — APPROVED** [review]

**Comment by yxxhero** [comment]
@sabre1041 hi. I'm sorry to bother you. could you have time to review this PR?

File: pkg/registry/util.go [hunk 1]
```diff
@@ -245,3 +246,13 @@ func addToMap(inputMap map[string]string, newKey string, newValue string) map[st
 	return inputMap
 
 }
+
+// See 2 (end of page 4) https://www.ietf.org/rfc/rfc2617.txt
+// "To receive authorization, the client sends the userid and password,
+// separated by a single colon (":") character, within a base64
+// encoded string in the credentials."
+// It is not meant to be urlencoded.
```
Comment (yxxhero) [inline pkg/registry/util.go:254]: Adding some unit tests will be better.

Comment (cuisongliu) [reply pkg/registry/util.go:254]: done

**Comment by sabre1041** [comment]
@cuisongliu just wanted to let you know that im currently teting

**Review by sabre1041 — APPROVED** [review]
LGTM

**Review by joejulian — APPROVED** [review]

**Pull request merged**